src/OpenTelemetry.AutoInstrumentation.StartupHook/RulesEngine/NativeProfilerDiagnosticsRule.cs

```diff
@@ -0,0 +1,122 @@
+
+    public NativeProfilerDiagnosticsRule()
+    {
+        Name = "Native profiler diagnoser";
+        Description = "Verifies that native profiler is correctly setup in case it's enabled.";
+    }
+
+    internal override bool Evaluate()
+    {
+        var isProfilerEnabled = EnvironmentHelper.GetEnvironmentVariable(ProfilerEnabledVariable) == "1";
+        if (!isProfilerEnabled)
+        {
+            Logger.Warning("{0} environment variable is not set to '1'. The CLR Profiler is disabled and no bytecode instrumentations are going to be injected.", ProfilerEnabledVariable);
+            return true;
+        }
+
+        try
+        {
+            if (NativeMethods.IsProfilerAttached())
+            {
+                return true;
+            }
+
+            Logger.Debug("IsProfilerAttached returned false.");
+        }
+        catch (Exception ex)
+        {
+            /* Native profiler is not attached. Continue with diagnosis */
+            Logger.Debug(ex, "Error checking if native profiler is attached.");
+        }
+
+        var profilerId = EnvironmentHelper.GetEnvironmentVariable(ProfilerIdVariable);
+        if (profilerId != ProfilerId)
+        {
+            Logger.Warning("The CLR profiler is enabled, but a different profiler ID was provided '{0}'.", profilerId);
+
+            // Different native profiler not assosiated to OTel might be used. We don't want to fail here.
+            return true;
+        }
+
+        if (Environment.Is64BitProcess)
+        {
+            VerifyPathVariables(Profiler64BitPathVariable, "64bit");
+        }
+        else
+        {
+            VerifyPathVariables(Profiler32BitPathVariable, "32bit");
+        }
+
+        return false;
+    }
+
+    private static void VerifyPathVariables(string archPathVariable, string expectedBitness)
+    {
+        if (TryPathVariable(archPathVariable, expectedBitness))
+        {
+            return;
+        }
+
+        if (TryPathVariable(ProfilerPathVariable, expectedBitness))
+        {
+            return;
+        }
+
+        Logger.Error("CLR profiler path is not defined. Define '{0}' or '{1}'.", ProfilerPathVariable, archPathVariable);
+    }
+
+    private static bool TryPathVariable(string profilerPathVariable, string expectedBitness)
+    {
+        var profilerPath = EnvironmentHelper.GetEnvironmentVariable(profilerPathVariable);
+
+        // Nothing to verify. Signal that VerifyVariables can continue searching for issues.
+        if (string.IsNullOrWhiteSpace(profilerPath))
+        {
+            return false;
+        }
+
+        if (File.Exists(profilerPath))
+        {
+            // File is found but profiler is not attaching.
+            Logger.Error("CLR profiler is not attaching profiler found at '{0}'. Recheck that {1} process is attaching {1} native profiler via {2}.", new object[] { profilerPath, expectedBitness, profilerPathVariable });
```

Comment:
Changing the message so it is correct if pinvoke fails or works but returns false. One could make a case that we should bail if pinvoke returns false, but, I think that in this unexpected case we may want the further diagnostic to help any possible investigation.

```suggestion
            Logger.Error("CLR profiler was not correctly loaded into the process. Profiler found at '{0}'. Recheck that {1} process is attaching {1} native profiler via {2}.", new object[] { profilerPath, expectedBitness, profilerPathVariable });
```

Reply:
https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/2897/commits/1cd8ba5c3c4b178e8a39491debc61d3eae7b7afd